ocaml/typing/printtyp.ml

```diff
@@ -2102,30 +2153,52 @@ and tree_of_functor_parameter = function
             Some (Ident.name id),
             Env.add_module ~arg:true id Mp_present ty_arg
       in
-      Some (name, tree_of_modtype ~ellipsis:false ty_arg), env
+      Some (name, tree_of_modtype ?abbrev ty_arg), env
 
-and tree_of_signature sg =
-  wrap_env (fun env -> env)(fun sg ->
-      let tree_groups = tree_of_signature_rec !printing_env sg in
-      List.concat_map (fun (_env,l) -> List.map snd l) tree_groups
-    ) sg
+and tree_of_signature ?abbrev = function
+  | [] -> []
+  | _ when Abbrev.exhausted abbrev -> [Osig_ellipsis]
+  | sg ->
+    Abbrev.deeper abbrev (fun () ->
+      wrap_env (fun env -> env)(fun sg ->
+        let max_items, trimmed = Abbrev.items abbrev sg in
+        let tree_groups = tree_of_signature_rec ?abbrev ?max_items !printing_env sg in
+        let items = List.concat_map (fun (_env,l) -> List.map snd l) tree_groups in
+        if trimmed then items @ [Osig_ellipsis] else items
+        ) sg
+    )
 
-and tree_of_signature_rec env' sg =
+and tree_of_signature_rec ?abbrev ?max_items env' sg =
   let structured = List.of_seq (Signature_group.seq sg) in
-  let collect_trees_of_rec_group group =
-    let env = !printing_env in
-    let env', group_trees =
-      Naming_context.with_ctx
-        (fun () -> trees_of_recursive_sigitem_group env group)
-    in
-    set_printing_env env';
-    (env, group_trees) in
+  let collect_trees_of_rec_group max_items group =
+    match max_items with
+    | Some n when n <= 0 -> (max_items, (!printing_env, []))
+    | Some _ | None ->
+        let env = !printing_env in
+        let env', group_trees =
+          Naming_context.with_ctx
+            (fun () -> trees_of_recursive_sigitem_group ?abbrev env group)
+        in
+        set_printing_env env';
+        let max_items, group_trees = match max_items with
+          | None -> None, group_trees
+          | Some n ->
+              let rec take n acc xs =
+                match n, xs with
+                | 0, _ | _, [] -> n, List.rev acc
+                | n, x :: xs -> take (n-1) (x :: acc) xs
+              in
+              let n, group_trees = take n [] group_trees in
+              Some n, group_trees
+        in
+        max_items, (env, group_trees)
+  in
   set_printing_env env';
-  List.map collect_trees_of_rec_group structured
+  snd (List.fold_left_map collect_trees_of_rec_group max_items structured)
```

Comment:
It looks to me like this could just use `Abbrev.items` to do its work, no? That is, instead of peeling off some allocation into `max_items` just to let that dwindle while iterating, peel off the items you need as you need them. That strikes me as simpler. Or maybe I'm missing something here?

Reply:
IIUC, what you suggest would expand things depth first. So given:
```
sig
  module M : sig <big> end
  <small>
end
```
we currently get
```
sig
  module M: sig <partial> ... end
  <small>
end
```
because we "pre-allocate" the items before descending. But if we didn't, we'd get
```
sig
  module M : sig <big> end
  ...
end
```
IMO the former is better (but shout if you disagree!). I added a comment.

Reply:
OK